Restore from restore plans for both backup types

soumyadeep2007 opened the pull request:
Iterate through timestamps found in restore plan of specified backup
and restore table data corresponding to each timestamp.
Also, ensure that if a restore is done on a legacy backup (having a
missing restore plan), a default restore plan is supplied. Here we
have assumed that such a legacy backup can only be a full backup.
Thus, the default restore plan would always be a single entry with
a list of all tables backed up in the full backup.

Co-authored-by: Chris Hajas <chajas@pivotal.io>
Co-authored-by: Soumyadeep Chakraborty <sochakraborty@pivotal.io>
Co-authored-by: Karen Huddleston <khuddleston@pivotal.io>

jmcatamney commented on restore/restore.go:
```diff
@@ -186,29 +193,39 @@ func restorePredata(metadataFilename string) {
-	filteredMasterDataEntries := globalTOC.GetDataEntriesMatching(*includeSchemas, *excludeSchemas, *includeRelations, *excludeRelations)
+	latestRestorePlan := backupConfig.RestorePlan
+	for i, fpInfo := range fpInfoList {
+		gplog.Info("Restoring data from backup with timestamp: %s	", fpInfo.Timestamp)
```
Any particular reason for the extra whitespace here?

soumyadeep2007 replied:
Typo! Thanks!

jmcatamney commented on restore/restore.go:
```diff
@@ -186,29 +193,39 @@ func restorePredata(metadataFilename string) {
 	}
-	gplog.Info("Restoring data")
-	filteredMasterDataEntries := globalTOC.GetDataEntriesMatching(*includeSchemas, *excludeSchemas, *includeRelations, *excludeRelations)
+	latestRestorePlan := backupConfig.RestorePlan
```
This might make more sense as just `restorePlan`, since it's not comparing multiple restore plans.

soumyadeep2007 replied:
We have multiple restore plans (one for each backup). At first glance it may not be apparent that the global config points to the latest config and by extension, the latest restore plan. That is why it is explicitly called out.

khuddlefish replied:
I'm good with latestRestorePlan. It is a bit more clear

khuddlefish commented on end_to_end/end_to_end_suite_test.go:
```diff
@@ -145,12 +145,24 @@ var _ = Describe("backup end to end integration tests", func() {
+			}
 		})
-		Context("Backup include filtering", func() {
+		Describe("Backup include filtering", func() {
```
What's the difference between Describe and Context for this purpose?

soumyadeep2007 replied:
Context is usually used to provide documentation about a particular scenario whereas Describe describes high level functionality that we want to test. In a unit test, Describe is typically a function, whereas here in end to end, it is a piece of higher level functionality.

That's why I replaced the context blocks.

khuddlefish commented on end_to_end/end_to_end_suite_test.go:
```diff
@@ -368,6 +380,32 @@ var _ = Describe("backup end to end integration tests", func() {
 			})
 		})
+		Describe("Incremental", func() {
+			It("restores from a full backup", func() {
```
This should already be encompassed in some other case. These tests take long enough that I would rather not add an unnecessary repeat case.

soumyadeep2007 replied:
Yeah, I will take that out.

chrishajas commented on utils/report.go:
```diff
@@ -42,6 +42,23 @@ type RestorePlanEntry struct {
 	TableFQNs []string
 }
 
+func EnsureDefaultRestorePlan(fpInfo *FilePathInfo, backupConfig *BackupConfig) {
```
We should add a test for this function.

soumyadeep2007 replied:
Cool, will add!

chrishajas commented on restore/wrappers.go:
```diff
@@ -62,6 +62,8 @@ func InitializeBackupConfig() {
 	utils.EnsureBackupVersionCompatibility(backupConfig.BackupVersion, version)
 	utils.EnsureDatabaseVersionCompatibility(backupConfig.DatabaseVersion, connectionPool.Version)
+
+	utils.EnsureDefaultRestorePlan(&globalFPInfo, backupConfig)
```
The above Ensure functions will exit the backup if the backup or database versions are incompatible, which differs from the purpose of `EnsureDefaultRestorePlan`. Let's name this function something more descriptive, such as `SetRestorePlanForLegacyBackup`.

soumyadeep2007 replied:
Makes sense.

chrishajas commented on utils/toc_test.go:
```diff
@@ -218,44 +218,62 @@ var _ = Describe("utils/toc tests", func() {
-	Context("GetDataEntriesMatching", func() {
-		It("returns matching entry on include schema", func() {
-			includeSchemas := []string{"schema1"}
+	Describe("GetDataEntriesMatching", func() {
```
Let's add tests for the restore plan (if the table matches the filter, but isn't in the restore plan for example, which would test a filtered restore case of incremental).

soumyadeep2007 replied:
The tests do test that. We have schema3.table3 which is not in the restore plan. For example in test : "returns matching entry on exclude table" schema3.table3 does match the filter but it is not in the restore plan.

chrishajas commented on end_to_end/test_tables_ddl.sql:
```diff
@@ -98,22 +98,30 @@ CREATE TABLE returns (
+
+CREATE TABLE ao2 (
+    i integer
+) WITH (appendonly=true);
```
Let's make one of these column oriented too. There's some logic in GPDB core (and possibly COPY) that differs between them.

soumyadeep2007 replied:
Sure!

chrishajas commented on end_to_end/test_tables_data.sql:
```diff
@@ -50,6 +50,10 @@ SET search_path = schema2, pg_catalog;
 
+INSERT INTO ao1 SELECT generate_series(1,50000);
+
+INSERT INTO ao2 SELECT generate_series(1,50000);
```
We already have a couple tables with 50k rows, so we can reduce this to like 1000 rows to make the tests a bit faster.

soumyadeep2007 replied:
Sure!

chrishajas commented on end_to_end/end_to_end_suite_test.go:
```diff
@@ -368,6 +380,23 @@ var _ = Describe("backup end to end integration tests", func() {
+				testhelper.AssertQueryRuns(backupConn, "INSERT into schema2.ao1 values(50001)")
+				defer testhelper.AssertQueryRuns(backupConn, "DELETE from schema2.ao1 where i=50001")
+
+				incrementalTimestamp := gpbackup(gpbackupPath, "--incremental", fullBackupTimestamp)
```
I think it would be good to have one more incremental in here, to ensure we handle incrementing on an incremental backup.

soumyadeep2007 replied:
Sure!

soumyadeep2007 commented:
@chrishajas Have a look. I added the tests, corrected the issue with filtering (double the no of tuples getting populated) and also refactored SetRestorePlanForLegacyBackup.

chrishajas commented on utils/toc.go:
```diff
@@ -176,6 +176,7 @@ func (toc *TOC) GetDataEntriesMatching(includeSchemas []string, excludeSchemas [
 	}
 
 	restorePlanTableSet := NewIncludeSet(restorePlanTableFQNs)
+	restorePlanTableSet.AlwaysMatchesFilter = false
```
Also would be good to have a unit test that catches this.

soumyadeep2007 replied:
Done.

chrishajas replied:
Just want to confirm that this was added. Once it's been added, looks good to merge it to the incremental branch!

soumyadeep2007 replied:
Yup done!
